### PR TITLE
JSON marshalable domain model for RFC006

### PR DESCRIPTION
**Pull request by reinkrul**

### CODE REVIEW

**Comment by woutslakhorst** [comment]
You can disable codecov by adding to `codecov.yml`:

```
github_checks:
  annotations: false
```

File: pkg/cert/certificate.go [hunk 1]
```diff
@@ -88,3 +90,26 @@ func (c NutsCertificate) getPartyIDFromSAN(oid asn1.ObjectIdentifier) (core.Part
 	}
 	return core.PartyID{}, nil
 }
+
+// MarshalJSON marshals the ASN.1 representation of the certificate as base64 JSON string.
+func (c NutsCertificate) MarshalJSON() ([]byte, error) {
```
Comment (woutslakhorst) [inline pkg/cert/certificate.go:95]: Missing tests for this file

File: pkg/nextgen/model/basetypes.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	"net/url"
+	"strings"
+	"unicode"
+)
+
+const oidURNNamespace = "oid"
+
+const errInvalidOIDURNFmt = "invalid OID URN: %s"
+
+type URL struct {
```
Comment (woutslakhorst) [inline pkg/nextgen/model/basetypes.go:16]: docs

File: pkg/nextgen/model/basetypes.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	url.URL
+}
+
+func NewURL(url url.URL) URL {
```
Comment (woutslakhorst) [inline pkg/nextgen/model/basetypes.go:20]: wow 5 times in a row

**Comment by reinkrul** [comment]
> You can disable codecov inline checks by adding to `codecov.yml`:
> 
> ```
> github_checks:
>   annotations: false
> ```

Well codecov is right in this isntance, I am missing some coverage there..

File: pkg/nextgen/model/basetypes.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// OIDURNValue is a URN in the OID scheme in the format of urn:oid:1.2.3.4:some-value
+type OIDURNValue struct {
+	urn   urn.URN
+	oid   string
+	value string
+}
+
+func NewOIDURN(oid string, value string) (OIDURNValue, error) {
```
Comment (woutslakhorst) [inline pkg/nextgen/model/basetypes.go:53]: docs, do I have to pass 'urn' and or 'oid' or nothing?

File: pkg/nextgen/model/model_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	"time"
+)
+
+var vendorId, _ = core.NewPartyID("1.2.3.4", "vendorId")
+var organizationId, _ = core.NewPartyID("4.3.1.1", "organizationId")
+var endpointType, _ = NewOIDURN("1.2.3.4", "http")
+var location, _ = url.Parse("https://nuts.nl")
+var notBefore = time.Date(2020, 10, 20, 20, 30, 0, 0, time.UTC)
+
+func TestMarshalVendor(t *testing.T) {
```
Comment (woutslakhorst) [inline pkg/nextgen/model/model_test.go:24]: What are you testing in this lib? The go language or the json library?

File: pkg/nextgen/model/model.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/nuts-foundation/nuts-registry/pkg/cert"
+	"github.com/nuts-foundation/nuts-registry/pkg/types"
+	"time"
+)
+
+type Proof interface {
+	// TODO
+}
+
+type Vendor struct {
```
Comment (woutslakhorst) [inline pkg/nextgen/model/model.go:14]: missing docs on types

Comment (woutslakhorst) [reply pkg/nextgen/model/model.go:14]: Also please refer to the right RFC

**Review by woutslakhorst — CHANGES_REQUESTED** [review]

File: pkg/nextgen/model/model_test.go (hunk 5 above)
Comment (reinkrul) [reply pkg/nextgen/model/model_test.go:24]: Conformance of the model to the spec (RFC006); whether the outputted JSON is like the spec has spec-ed it (correctness of tags and types).

**Comment by reinkrul** [comment]
What do you think about wrapping built-in types (like url.URL) just for JSON marshalling? The advantage is you get type safety but the disadvantage is we produce quite a bit of code and quirky API's. You think it's worth it?

**Comment by woutslakhorst** [comment]
> What do you think about wrapping built-in types (like url.URL) just for JSON marshalling? The advantage is you get type safety but the disadvantage is we produce quite a bit of code and quirky API's. You think it's worth it?

at least this centralizes the problem to a single file ;). We'll just have to be weary in the future.

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
Could you enable code-climate in `.codeclimate.yml`:

```
plugins:
  gofmt:
    enabled: true
  golint:
    enabled: true
  govet:
    enabled: true
```

If codeclimate is all green, then ok to merge.

**Comment by reinkrul** [comment]
Not going to fix the whole world as part of this PR, something for the "rustige uurtjes".

**Pull request merged**